File: 665.py
class Solution:
    def checkPossibility(self, nums: list[int]) -> bool:
        changed = False
        for i in range(len(nums) - 1):
            if nums[i] <= nums[i + 1]:
                continue
            else:
                if changed:
                    return False
                if i > 0 and nums[i + 1] < nums[i - 1]:
                    nums[i + 1] = nums[i]
                else:
                    nums[i] = nums[i + 1]
                changed = True
        return True


s = Solution()
assert s.checkPossibility([3, 4, 2, 3]) == False
assert s.checkPossibility([4, 2, 3]) == True
assert s.checkPossibility([4, 2, 1]) == False
assert s.checkPossibility([5, 7, 1, 8]) == True

File: 665.cpp
#include <vector>
using namespace std;

class Solution {
public:
  bool checkPossibility(vector<int> &nums) {
    bool changed = false;
    for (int i = 0; i < nums.size(); ++i) {
      int num = nums[i];
      if (i == nums.size() - 1 || num <= nums[i + 1]) {
        continue;
      }
      if (changed) {
        return false;
      }
      changed = true;
      if (i == 0 || nums[i - 1] <= nums[i + 1]) {
        nums[i] = nums[i + 1];
      } else {
        nums[i + 1] = nums[i];
      }
    }
    return true;
  }
};
